File: jbake-core/src/test/java/org/jbake/app/AsciidocParserTest.java
package org.jbake.app;

import java.io.File;
import java.io.PrintWriter;
import java.nio.file.Path;

import org.jbake.TestUtils;
import org.jbake.app.configuration.ConfigUtil;
import org.jbake.app.configuration.DefaultJBakeConfiguration;
import org.jbake.model.DocumentModel;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import static org.assertj.core.api.Assertions.assertThat;
import static org.jbake.app.configuration.PropertyList.ASCIIDOCTOR_ATTRIBUTES;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;

class AsciidocParserTest {

    @TempDir
    private Path folder;

    private DefaultJBakeConfiguration config;
    private Parser parser;
    private File rootPath;

    private File asciidocWithSource;
    private File validAsciidocFile;
    private File invalidAsciiDocFile;
    private File validAsciiDocFileWithoutHeader;
    private File invalidAsciiDocFileWithoutHeader;
    private File validAsciiDocFileWithHeaderInContent;
    private File validAsciiDocFileWithoutJBakeMetaData;

    private String validHeader = "title=This is a Title = This is a valid Title\nstatus=draft\ntype=post\ndate=2013-09-02\n~~~~~~";
    private String invalidHeader = "title=This is a Title\n~~~~~~";

    @BeforeEach
    void createSampleFile() throws Exception {
        rootPath = TestUtils.getTestResourcesAsSourceFolder();
        config = (DefaultJBakeConfiguration) new ConfigUtil().loadConfig(rootPath);
        parser = new Parser(config);

        asciidocWithSource = folder.resolve("asciidoc-with-source.ad").toFile();
        PrintWriter out = new PrintWriter(asciidocWithSource);
        out.println(validHeader);
        out.println("= Hello, AsciiDoc!");
        out.println("Test User <user@test.org>");
        out.println("");
        out.println("JBake now supports AsciiDoc.");
        out.println("");
        out.println("```");
        out.println("#!/bin/bash");
        out.println("");
        out.println("echo 'hello world!'");
        out.println("```");
        out.println("");
        out.println("{testattribute}");

        out.close();

        validAsciidocFile = folder.resolve("valid.ad").toFile();
        out = new PrintWriter(validAsciidocFile);
        out.println(validHeader);
        out.println("= Hello, AsciiDoc!");
        out.println("Test User <user@test.org>");
        out.println("");
        out.println("JBake now supports AsciiDoc.");
        out.close();

        invalidAsciiDocFile = folder.resolve("invalid.ad").toFile();
        out = new PrintWriter(invalidAsciiDocFile);
        out.println(invalidHeader);
        out.println("= Hello, AsciiDoc!");
        out.println("Test User <user@test.org>");
        out.println("");
        out.println("JBake now supports AsciiDoc.");
        out.close();

        validAsciiDocFileWithoutHeader = folder.resolve("validwoheader.ad").toFile();
        out = new PrintWriter(validAsciiDocFileWithoutHeader);
        out.println("= Hello: AsciiDoc!");
        out.println("Test User <user@test.org>");
        out.println("2013-09-02");
        out.println(":jbake-status: published");
        out.println(":jbake-type: page");
        out.println("");
        out.println("JBake now supports AsciiDoc.");
        out.close();

        invalidAsciiDocFileWithoutHeader = folder.resolve("invalidwoheader.ad").toFile();
        out = new PrintWriter(invalidAsciiDocFileWithoutHeader);
        out.println("= Hello, AsciiDoc!");
        out.println("Test User <user@test.org>");
        out.println("2013-09-02");
        out.println(":jbake-status: published");
        out.println("");
        out.println("JBake now supports AsciiDoc.");
        out.close();

        validAsciiDocFileWithHeaderInContent = folder.resolve("validheaderincontent.ad").toFile();
        out = new PrintWriter(validAsciiDocFileWithHeaderInContent);
        out.println("= Hello, AsciiDoc!");
        out.println("Test User <user@test.org>");
        out.println("2013-09-02");
        out.println(":jbake-status: published");
        out.println(":jbake-type: page");
        out.println("");
        out.println("JBake now supports AsciiDoc.");
        out.println("");
        out.println("----");
        out.println("title=Example Header");
        out.println("date=2013-02-01");
        out.println("type=post");
        out.println("tags=tag1, tag2");
        out.println("status=published");
        out.println("~~~~~~");
        out.println("----");
        out.close();

        validAsciiDocFileWithoutJBakeMetaData = folder.resolve("validwojbakemetadata.ad").toFile();
        out = new PrintWriter(validAsciiDocFileWithoutJBakeMetaData);
        out.println("= Hello: AsciiDoc!");
        out.println("Test User <user@test.org>");
        out.println("2013-09-02");
        out.println("");
        out.println("JBake now supports AsciiDoc documents without JBake meta data.");
        out.close();
    }


    @Test
    void parseAsciidocFileWithPrettifyAttribute() {

        config.setProperty(ASCIIDOCTOR_ATTRIBUTES.getKey(), "source-highlighter=prettify");
        DocumentModel map = parser.processFile(asciidocWithSource);
        assertNotNull(map);
        assertEquals("draft", map.getStatus());
        assertEquals("post", map.getType());
        assertThat(map.getBody())
            .contains("class=\"paragraph\"")
            .contains("<p>JBake now supports AsciiDoc.</p>")
            .contains("class=\"prettyprint highlight\"");

        assertThat(map.getBody()).doesNotContain("I Love Jbake");
        System.out.println(map.getBody());
    }

    @Test
    void parseAsciidocFileWithCustomAttribute() {

        config.setProperty(ASCIIDOCTOR_ATTRIBUTES.getKey(), "source-highlighter=prettify,testattribute=I Love Jbake");
        DocumentModel map = parser.processFile(asciidocWithSource);
        assertNotNull(map);
        assertEquals("draft", map.getStatus());
        assertEquals("post", map.getType());
        assertThat(map.getBody())
            .contains("I Love Jbake")
            .contains("class=\"prettyprint highlight\"");

        System.out.println(map.getBody());
    }

    @Test
    void parseValidAsciiDocFile() {
        DocumentModel map = parser.processFile(validAsciidocFile);
        assertNotNull(map);
        assertEquals("draft", map.getStatus());
        assertEquals("post", map.getType());
        assertThat(map.getBody())
            .contains("class=\"paragraph\"")
            .contains("<p>JBake now supports AsciiDoc.</p>");
    }

    @Test
    void parseInvalidAsciiDocFile() {
        DocumentModel map = parser.processFile(invalidAsciiDocFile);
        assertNull(map);
    }

    @Test
    void parseValidAsciiDocFileWithoutHeader() {
        DocumentModel map = parser.processFile(validAsciiDocFileWithoutHeader);
        assertNotNull(map);
        assertEquals("Hello: AsciiDoc!", map.get("title"));
        assertEquals("published", map.getStatus());
        assertEquals("page", map.getType());
        assertThat(map.getBody())
            .contains("class=\"paragraph\"")
            .contains("<p>JBake now supports AsciiDoc.</p>");
    }

    @Test
    void parseInvalidAsciiDocFileWithoutHeader() {
        DocumentModel map = parser.processFile(invalidAsciiDocFileWithoutHeader);
        assertNull(map);
    }

    @Test
    void parseValidAsciiDocFileWithExampleHeaderInContent() {
        DocumentModel map = parser.processFile(validAsciiDocFileWithHeaderInContent);
        assertNotNull(map);
        assertEquals("published", map.getStatus());
        assertEquals("page", map.getType());
        assertThat(map.getBody())
            .contains("class=\"paragraph\"")
            .contains("<p>JBake now supports AsciiDoc.</p>")
            .contains("class=\"listingblock\"")
            .contains("class=\"content\"")
            .contains("<pre>")
            .contains("title=Example Header")
            .contains("date=2013-02-01")
            .contains("tags=tag1, tag2");
    }

    @Test
    void parseValidAsciiDocFileWithoutJBakeMetaDataUsingDefaultTypeAndStatus() {
        config.setDefaultStatus("published");
        config.setDefaultType("page");
        Parser parser = new Parser(config);
        DocumentModel map = parser.processFile(validAsciiDocFileWithoutJBakeMetaData);
        assertNotNull(map);
        assertEquals("published", map.getStatus());
        assertEquals("page", map.getType());
        assertThat(map.getBody())
            .contains("<p>JBake now supports AsciiDoc documents without JBake meta data.</p>");
    }

}

File: jbake-core/src/test/java/org/jbake/app/AsciidocParserTest.kt
package org.jbake.app

import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.nulls.shouldBeNull
import io.kotest.matchers.nulls.shouldNotBeNull
import io.kotest.matchers.shouldBe
import io.kotest.matchers.string.shouldContain
import io.kotest.matchers.string.shouldNotContain
import org.jbake.TestUtils
import org.jbake.app.configuration.ConfigUtil
import org.jbake.app.configuration.DefaultJBakeConfiguration
import org.jbake.app.configuration.PropertyList.ASCIIDOCTOR_ATTRIBUTES
import org.jbake.model.DocumentModel
import org.jbake.parser.Parser
import java.io.File

class AsciidocParserTest : StringSpec({

    val validHeader = """
        |title=This is a Title = This is a valid Title
        |status=draft
        |type=post
        |date=2013-09-02
        |~~~~~~
        """.trimMargin()

    val invalidHeader = """
        |title=This is a Title
        |~~~~~~
        """.trimMargin()

    val asciidocWithSourceContent = validHeader + "\n" + """
        |= Hello, AsciiDoc!
        |Test User <user@test.org>
        |
        |JBake now supports AsciiDoc.
        |
        |```
        |#!/bin/bash
        |
        |echo 'hello world!'
        |```
        |
        |{testattribute}
        """.trimMargin()

    val validAsciidocContent = validHeader + "\n" + """
        |= Hello, AsciiDoc!
        |Test User <user@test.org>
        |
        |JBake now supports AsciiDoc.
        """.trimMargin()

    val invalidAsciiDocContent = invalidHeader + "\n" + """
        |= Hello, AsciiDoc!
        |Test User <user@test.org>
        |
        |JBake now supports AsciiDoc.
        """.trimMargin()

    val validAsciiDocWithoutHeaderContent = """
        |= Hello: AsciiDoc!
        |Test User <user@test.org>
        |2013-09-02
        |:jbake-status: published
        |:jbake-type: page
        |
        |JBake now supports AsciiDoc.
        """.trimMargin()

    val invalidAsciiDocWithoutHeaderContent = """
        |= Hello, AsciiDoc!
        |Test User <user@test.org>
        |2013-09-02
        |:jbake-status: published
        |
        |JBake now supports AsciiDoc.
        """.trimMargin()

    val validAsciiDocWithHeaderInContentContent = """
        |= Hello, AsciiDoc!
        |Test User <user@test.org>
        |2013-09-02
        |:jbake-status: published
        |:jbake-type: page
        |
        |JBake now supports AsciiDoc.
        |
        |----
        |title=Example Header
        |date=2013-02-01
        |type=post
        |tags=tag1, tag2
        |status=published
        |~~~~~~
        |----
        """.trimMargin()

    val validAsciiDocWithoutJBakeMetaDataContent = """
        |= Hello: AsciiDoc!
        |Test User <user@test.org>
        |2013-09-02
        |
        |JBake now supports AsciiDoc documents without JBake meta data.
        """.trimMargin()
    
    fun assertBasicAsciiDocContent(body: String) {
        body shouldContain "class=\"paragraph\""
        body shouldContain "<p>JBake now supports AsciiDoc.</p>"
    }

    fun assertMetadata(map: DocumentModel?, status: String, type: String) {
        map.shouldNotBeNull()
        map.status shouldBe status
        map.type shouldBe type
    }

    lateinit var tempDir: File
    lateinit var rootPath: File
    lateinit var config: DefaultJBakeConfiguration
    lateinit var parser: Parser
    
    lateinit var asciidocWithSource: File
    lateinit var validAsciidocFile: File
    lateinit var invalidAsciiDocFile: File
    lateinit var validAsciiDocFileWithoutHeader: File
    lateinit var invalidAsciiDocFileWithoutHeader: File
    lateinit var validAsciiDocFileWithHeaderInContent: File
    lateinit var validAsciiDocFileWithoutJBakeMetaData: File

    fun createTestFile(name: String, content: String) =
        tempDir.resolve(name).apply { createNewFile(); writeText(content) }

    beforeTest {
        tempDir = java.nio.file.Files.createTempDirectory("jbake-test").toFile()
        rootPath = TestUtils.testResourcesAsSourceDir
        config = ConfigUtil().loadConfig(rootPath) as DefaultJBakeConfiguration
        parser = Parser(config)
        
        asciidocWithSource = createTestFile("asciidoc-with-source.ad", asciidocWithSourceContent)
        validAsciidocFile = createTestFile("valid.ad", validAsciidocContent)
        invalidAsciiDocFile = createTestFile("invalid.ad", invalidAsciiDocContent)
        validAsciiDocFileWithoutHeader = createTestFile("validwoheader.ad", validAsciiDocWithoutHeaderContent)
        invalidAsciiDocFileWithoutHeader = createTestFile("invalidwoheader.ad", invalidAsciiDocWithoutHeaderContent)
        validAsciiDocFileWithHeaderInContent = createTestFile("validheaderincontent.ad", validAsciiDocWithHeaderInContentContent)
        validAsciiDocFileWithoutJBakeMetaData = createTestFile("validwojbakemetadata.ad", validAsciiDocWithoutJBakeMetaDataContent)
    }


    "parseAsciidocFileWithPrettifyAttribute" {
        config.setProperty(ASCIIDOCTOR_ATTRIBUTES.key, "source-highlighter=prettify")
        val map = parser.processFile(asciidocWithSource)
        assertMetadata(map, "draft", "post")
        assertBasicAsciiDocContent(map!!.body)
        map.body shouldContain "class=\"prettyprint highlight\""
        map.body shouldNotContain "I Love Jbake"
    }

    "parseAsciidocFileWithCustomAttribute" {
        config.setProperty(ASCIIDOCTOR_ATTRIBUTES.key, "source-highlighter=prettify,testattribute=I Love Jbake")
        val map = parser.processFile(asciidocWithSource)
        assertMetadata(map, "draft", "post")
        map!!.body shouldContain "I Love Jbake"
        map.body shouldContain "class=\"prettyprint highlight\""
    }

    "parseValidAsciiDocFile" {
        val map = parser.processFile(validAsciidocFile)
        assertMetadata(map, "draft", "post")
        assertBasicAsciiDocContent(map!!.body)
    }

    "parseInvalidAsciiDocFile" {
        val map = parser.processFile(invalidAsciiDocFile)
        map.shouldBeNull()
    }

    "parseValidAsciiDocFileWithoutHeader" {
        val map = parser.processFile(validAsciiDocFileWithoutHeader)
        assertMetadata(map, "published", "page")
        map!!["title"] shouldBe "Hello: AsciiDoc!"
        assertBasicAsciiDocContent(map.body)
    }

    "parseInvalidAsciiDocFileWithoutHeader" {
        val map = parser.processFile(invalidAsciiDocFileWithoutHeader)
        map.shouldBeNull()
    }

    "parseValidAsciiDocFileWithExampleHeaderInContent" {
        val map = parser.processFile(validAsciiDocFileWithHeaderInContent)
        assertMetadata(map, "published", "page")
        assertBasicAsciiDocContent(map!!.body)
        map.body shouldContain "class=\"listingblock\""
        map.body shouldContain "class=\"content\""
        map.body shouldContain "<pre>"
        map.body shouldContain "title=Example Header"
        map.body shouldContain "date=2013-02-01"
        map.body shouldContain "tags=tag1, tag2"
    }

    "parseValidAsciiDocFileWithoutJBakeMetaDataUsingDefaultTypeAndStatus" {
        config.setDefaultStatus("published")
        config.setDefaultType("page")
        val parser = Parser(config)
        val map = parser.processFile(validAsciiDocFileWithoutJBakeMetaData)
        assertMetadata(map, "published", "page")
        map!!.body shouldContain "<p>JBake now supports AsciiDoc documents without JBake meta data.</p>"
    }


    afterTest {
        tempDir.deleteRecursively()
    }
})
